### PR TITLE
Try uncontrol style

### PR DESCRIPTION
**Pull request by minhnhut16**

### CODE REVIEW

File: src/components/TreeViewer/hooks/useTreeViewer.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import { useRef } from 'react';
+
+export default function useTreeViewer() {
+  const treeRef = useRef(null);
+  return { treeRef, ...treeRef?.current };
```
Comment (ghost) [inline src/components/TreeViewer/hooks/useTreeViewer.js]: what api you expose of this hook?

File: src/components/TreeViewer/hooks/useExpand.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  const togleExpandedNode = useCallback(
+    nodeId => {
+      if (expandedNodes.includes(nodeId)) {
+        setExpandedNodes(expandedNodes.filter(id => id !== nodeId));
+      } else {
+        setExpandedNodes([...expandedNodes, nodeId]);
+      }
+    },
+    [expandedNodes]
```
Comment (ghost) [inline src/components/TreeViewer/hooks/useExpand.js]: this should be use state functional update like `setState(prev => ... )`, you can avoid `expandedNodes` in deps and avoid re-create function

File: src/components/MultilevelSelect/index.js [hunk 3]
```diff
@@ -1,50 +1,121 @@
...
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+  justify-content: center;
+  height: 100%;
+`;
+const fakeApi = () => {};
+
+const MultilevelSelect = ({ defaultTreeData, open, onCancel, onFinish }) => {
+  const treeInsRef = useRef(new Tree(defaultTreeData));
```
Comment (ghost) [inline src/components/MultilevelSelect/index.js]: I think `const treeInsRef = useRef(new Tree(defaultTreeData)).current;` is enough

File: src/components/TreeViewer/index.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      }
+    },
+    [onChange, selectedNode.id, togleExpandedNode, treeIns]
+  );
+
+  const renderedTree = useMemo(() => {
+    const Node = CustomNode ?? DefaultNode;
+
+    if (treeIns?.renderTree) {
+      return treeIns?.renderTree({
```
Comment (ghost) [inline src/components/TreeViewer/index.js]: remove check, you have condition above

File: src/components/TreeViewer/hooks/useExpand.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import { useCallback, useState } from 'react';
+
+export default function useExpand() {
+  const [expandedNodes, setExpandedNodes] = useState([]);
+
+  const togleExpandedNode = useCallback(nodeId => {
```
Comment (ghost) [inline src/components/TreeViewer/hooks/useExpand.js]: typo should be `toggleExpandedNode`

File: src/components/TreeViewer/index.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        });
+      }
+    },
+    [onChange, selectedNode.id, togleExpandedNode, treeIns]
+  );
+
+  const renderedTree = useMemo(() => {
+    const Node = CustomNode ?? DefaultNode;
+
+    if (treeIns?.renderTree) {
```
Comment (ghost) [inline src/components/TreeViewer/index.js]: I think you should check and return null soon, should be

```js
if (!treeIns?.renderTree) return null;

// ... your logic render
```

Comment (an-nguyen-2-ts) [reply src/components/TreeViewer/index.js:34]: I think you should check when `renderTree` dont have first, should be
```js
if (!treeIns?.renderTree) return null;

// ... your logic render
```

File: src/entities/Tree.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+  constructor(treeData) {
+    this.treeData = treeData;
+    this.traversalByNode = this.traversalByNode.bind(this);
+    this.updateNode = this.updateNode.bind(this);
+    this.renderTree = this.renderTree.bind(this);
+  }
+
+  traversalByNode(curNode, fn) {
+    const c = this;
```
Comment (ghost) [inline src/entities/Tree.js]: We should use this name, you mean what? "C" in Vietnamese we can understand in many means

Comment (ghost) [reply src/entities/Tree.js]: I think should be `ctx` stands for `context`

File: src/utils.js [hunk 8]
```diff
@@ -0,0 +1,3 @@
+export function isNotEmptyArray(arr) {
+  return Array.isArray(arr) && arr.length;
```
Comment (ghost) [inline src/utils.js]: why we don't implement a method name 
isEmptyArray

File: src/entities/Tree.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
+import cloneDeep from 'lodash/cloneDeep';
+
+export class Tree {
+  treeData;
+
+  constructor(treeData) {
+    this.treeData = treeData;
+    this.traversalByNode = this.traversalByNode.bind(this);
+    this.updateNode = this.updateNode.bind(this);
```
Comment (ghost) [inline src/entities/Tree.js]: you don't need to write it if your method is arrow function

File: src/components/TreeViewer/index.js [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+  const [selectedNode, setSelectedNode] = useState({});
+  const forceRender = useForceRender();
+
+  const { treeData, traversalByNode, renderTree } = treeIns || {};
+  const Node = CustomNode ?? DefaultNode;
+
+  const handleClick = useCallback(
+    curNode => {
+      if (traversalByNode) {
+        traversalByNode(curNode, () => {
```
Comment (ghost) [inline src/components/TreeViewer/index.js]: don't need traversal here

File: src/entities/Tree.js [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+  updateNode(curNode, onUpdate) {
+    const c = this;
+    const clonedTree = cloneDeep(this.treeData);
+
+    function traversal(node) {
+      if (node.id === curNode.id) {
+        Object.assign(node, cloneDeep(curNode));
+
+        if (onUpdate) {
```
Comment (ghost) [inline src/entities/Tree.js]: remove

**Pull request merged**